### PR TITLE
Fix compiler warning reported for target-typed new under collection initializer

### PR DESCRIPTION
**Pull request by alrz**
Fixes https://github.com/dotnet/roslyn/issues/60960

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/ObsoleteAttributeHelpers.cs [hunk 1]
```diff
@@ -157,7 +157,8 @@ internal static DiagnosticInfo CreateObsoleteDiagnostic(Symbol symbol, BinderFla
             }
 
             // Issue a specialized diagnostic for add methods of collection initializers
-            var isColInit = location.Includes(BinderFlags.CollectionInitializerAddMethod);
+            var isColInit = location.Includes(BinderFlags.CollectionInitializerAddMethod) &&
+                            symbol.Name == WellKnownMemberNames.CollectionInitializerAddMethodName;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/ObsoleteAttributeHelpers.cs:161]: >symbol.Name == WellKnownMemberNames.CollectionInitializerAddMethodName [](http://example.com/codeflow?start=28&length=70)

I think we should make sure that we are getting "correct" BinderFlags value instead. The value that we would get if we were using an object expression with explicit type. Probably we should save the Binder in ```BoundUnconvertedObjectCreationExpression``` node and use that binder to bind "conversion" to the target type (in ```ConvertObjectCreationExpression```).

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

**Comment by alrz** [comment]
(had to force-push because on main `build` look for dotnet and fails even though I have it installed)

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 2]
```diff
@@ -447,13 +448,11 @@ private BoundExpression ConvertSwitchExpression(BoundUnconvertedSwitchExpression
...
                 var oldCase = source.SwitchArms[i];
-                Debug.Assert(oldCase.Syntax is SwitchExpressionArmSyntax);
-                var binder = GetRequiredBinder(oldCase.Syntax);
                 var oldValue = oldCase.Value;
                 var newValue =
                     targetTyped
-                    ? binder.CreateConversion(oldValue.Syntax, oldValue, underlyingConversions[i], isCast: false, conversionGroupOpt: null, destination, diagnostics)
-                    : binder.GenerateConversionForAssignment(destination, oldValue, diagnostics);
+                    ? CreateConversion(oldValue.Syntax, oldValue, underlyingConversions[i], isCast: false, conversionGroupOpt: null, destination, diagnostics)
+                    : GenerateConversionForAssignment(destination, oldValue, diagnostics);
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:455]: Reverts https://github.com/dotnet/roslyn/pull/54234 - no longer needed.

**Comment by AlekseyTs** [comment]
>         private BoundExpression ConvertObjectCreationExpression(

I think we want to make this function static too. Basically for any context information it should use the binder from the node.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:346 in e26f722. [](commit_id = e26f722662d16e070349ee7352a0a6939f351f39, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 3]
```diff
@@ -462,13 +463,11 @@ private BoundExpression ConvertSwitchExpression(BoundUnconvertedSwitchExpression
             for (int i = 0, n = source.SwitchArms.Length; i < n; i++)
             {
                 var oldCase = source.SwitchArms[i];
-                Debug.Assert(oldCase.Syntax is SwitchExpressionArmSyntax);
-                var binder = GetRequiredBinder(oldCase.Syntax);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:466]: >var binder = GetRequiredBinder(oldCase.Syntax); [](http://example.com/codeflow?start=16&length=47)

I guess reverting this change feels fine, but we probably need to open an issue to follow up on other target-typed expressions to make sure they are also using the "right" binder for converting to the target type.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 3 above)
Comment (alrz) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:466]: Afaict this is the only place that needed this adjustment as mentioned in the linked PR. Now that we're using the original binder we don't need to do any of that.

**Comment by alrz** [comment]
>I think we want to make this function static too. Basically for any context information it should use the binder from the node.

CheckOverflowAtRuntime (line 355) is an instance member.

**Comment by AlekseyTs** [comment]
>> I think we want to make this function static too. Basically for any context information it should use the binder from the node.

> CheckOverflowAtRuntime (line 355) is an instance member.

Exactly my point. 

---
In reply to: [1116409480](https://github.com/dotnet/roslyn/pull/61040#issuecomment-1116409480) [](http://example.com/codeflow?ancestors=1116409480,1116409480)

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:466]: > Now that we're using the original binder we don't need to do any of that.

This only covers target-typed new. I am talking about "an issue to follow up on **other** target-typed expressions"

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 8)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 3 above)
Comment (alrz) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:466]: https://github.com/dotnet/roslyn/issues/61179

**Review by 333fred — APPROVED** [review]

**Comment by 333fred** [comment]
Thanks @alrz!

**Pull request merged**